src/Servers/Kestrel/Core/src/Internal/Http2/Http2KeepAlive.cs

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2
+{
+    internal enum KeepAliveState
+    {
+        None,
+        SendPing,
+        PingSent,
+        Timeout
+    }
+
+    internal class Http2KeepAlive
+    {
+        // An empty ping payload
+        internal static readonly ReadOnlySequence<byte> PingPayload = new ReadOnlySequence<byte>(new byte[8]);
+
+        private readonly object _lock = new object();
```

Comment:
I know I said we should check for timeouts in Tick, but looking at this now, I think we should do what we do in UpdateCompletedStreams and check fort timeouts in ProcessKeepAlive when necessary. ProcessKeepAlive is called at least once a second since Tick calls CancelPendingRead.

I was trying to avoid unnecessary calls to DateTimeOffset.UtcNow, but HeartbeatManager.UtcNowTicks is makes a cached value cheaply available to us.

By doing this, we can get rid of the lock, get rid of Tick, and get rid of PingSent since we can set _pingSentTimestamp in ProcessKeepAlive when it returns KeepAliveState.SendPing and know that a ping will have been sent by the time ProcessKeepAlive is called again.

Reply:
Done. I've kept SendPing and PingSent status to keep state simple.